src/main/scala/org/constellation/primitives/TransactionGenerator.scala

```diff
@@ -128,21 +135,38 @@ class TransactionGenerator[F[_]: Concurrent](
     addresses: Seq[(String, Option[AddressCacheData])],
     peers: Seq[(Id, PeerData)]
   ): F[Transaction] =
-    transactionService
-      .createTransaction(dao.selfAddressStr, randomAddressFrom(addresses), randomAmount(rangeAmount), dao.keyPair)
+    transactionService.transactionChainService.getLastTransactionRef(dao.selfAddressStr)
+      .flatMap { ref: LastTransactionRef =>
+        transactionService
+          .createTransaction(
```

Comment:
this will not work because of concurrency. the read and set must be done in one step (`modify` or `update`) on the `Ref`. During `modify` or `update`, the final value will be set if and only if the ref didn't change during the update process. With `.get.flatMap(... set)` you are getting the value, then something is able to change this value and you will set already outdated value as a new one, which will break the tx chain

Reply:
Actually you can see the implementation of `Ref`:

```
 private final class SyncRef[F[_], A](ar: AtomicReference[A])(implicit F: Sync[F]) extends Ref[F, A] {

    def get: F[A] = F.delay(ar.get)

    def set(a: A): F[Unit] = F.delay(ar.set(a))

...

    def update(f: A => A): F[Unit] =
      modify(a => (f(a), ()))

    def modify[B](f: A => (A, B)): F[B] = {
      @tailrec
      def spin: B = {
        val c = ar.get
        val (u, b) = f(c)
        if (!ar.compareAndSet(c, u)) spin
        else b
      }
      F.delay(spin)
    }

...
```